### PR TITLE
Implementation of makeDate and makeDate32

### LINKED ISSUE #30895: Functions `makeDate`, `makeDateTime`, `YYYYMMDDToDate`, `changeYear` and similar.
**Describe solution you'd like:**

```
makeDate(year, month, day)
makeDate32(year, month, day)
makeDateTime(year, month, day, hour, minute, second, [timezone])
makeDateTime64(year, month, day, hour, minute, second, [resolution], [fraction], [timezone])

YYYYMMDDToDate(num)
YYYYMMDDToDate32(num)
YYYYMMDDhhmmssToDateTime(num, [timezone])
YYYYMMDDhhmmssToDateTime64(num, [resolution], [timezone]) -- no fraction arg, it's implicitly extracted from floating point or decimal num.

-- returns the same data type as dt:
changeYear(dt, new_value)
changeMonth(dt, new_value)
changeDay(dt, new_value)

-- if argument is Date - returns DateTime, if Date32, returns DateTime64:
changeHour(dt, new_value)
changeMinute(dt, new_value)
changeSecond(dt, new_value)
```

### PR DESCRIPTION
**Pull request by davenger**
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added functions makeDate(year, month, day), makeDate32(year, month, day)

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Functions/makeDate.cpp [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        }
+
+        return std::make_shared<typename Traits::ReturnDataType>();
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    {
+        const DataTypePtr converted_argument_type = std::make_shared<DataTypeFloat64>();
+        Columns converted_arguments;
```
Comment (alesapin) [inline src/Functions/makeDate.cpp:81]: reserve `argument.size()`?

File: src/Functions/makeDate.cpp [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+            }
+
+            result_data[i] = day_num;
+        }
+
+        return res_column;
+    }
+
+private:
+    FunctionToFloat64 convert_to_number_func;
```
Comment (alesapin) [inline src/Functions/makeDate.cpp:132]: Do we really need Float64? What will happen if we replace it with Float32 or even Int16 (at least overflow issues)?

File: src/Functions/makeDate.cpp [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+namespace
+{
+
+// A helper function to simplify comparisons of valid YYYY-MM-DD values for <,>,=
+inline constexpr Int64 YearMonthDayToSingleInt(Int64 year, Int64 month, Int64 day)
+{
+    return year*512 + month*32 + day;
```
Comment (alesapin) [inline src/Functions/makeDate.cpp:29]:
```suggestion
    return year * 512 + month * 32 + day;
```

File: tests/queries/0_stateless/02242_make_date.sql [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+select makeDate(1969,1,1);
+select makeDate(1969,12,1);
+select makeDate(1969,12,31);
+select makeDate(1970,0,0);
+select makeDate(1970,0,1);
+select makeDate(1970,1,0);
+select makeDate(1990,0,1);
+select makeDate(1990,1,0);
+
+select makeDate('1980', '10', '20');
```
Comment (alesapin) [inline tests/queries/0_stateless/02242_make_date.sql:47]: So now even something like `makeDate('1980.05', '10', '20.3')` should work?

**Review by alesapin — COMMENTED** [review]
Actually I'm not sure that we have to support all data types as arguments... Maybe leave only integers  Everything will be faster and simpler :)

File: src/Functions/makeDate.cpp [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            result_data[i] = day_num;
+        }
+
+        return res_column;
+    }
+
+private:
+    FunctionToFloat64 convert_to_number_func;
+    FunctionToFloat64OrZero parse_number_func;
+    FunctionOr or_func;
```
[the remark this answers is not in the document]
Comment (davenger) [reply src/Functions/makeDate.cpp]: removed unused

File: tests/queries/0_stateless/02242_make_date.sql (hunk 4 above)
Comment (davenger) [reply tests/queries/0_stateless/02242_make_date.sql]: I replaced my custom to conversion logic to castColumn() that is used in other similar cases.
Now string arguments will not be accepted, only numbers will be allowed.

File: src/Functions/makeDate.cpp [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        Columns converted_arguments;
+        for (size_t i = 0; i < arguments.size(); ++i)
+        {
+            const auto & argument_type = arguments[i].type;
+            ColumnPtr argument_column = arguments[i].column->convertToFullColumnIfConst();
+
+            if (converted_argument_type->equals(*argument_type))
+                converted_arguments.push_back(argument_column);
+            else if (isNumber(argument_type))
+                converted_arguments.push_back(convert_to_number_func.executeImpl(
```
[the remark this answers is not in the document]
Comment (davenger) [reply src/Functions/makeDate.cpp]: replaced my custom to conversion logic to castColumn() that is used in other similar cases

**Review by alesapin — APPROVED** [review]

**Comment by alesapin** [comment]
Let's add it to category "New feature" and add Changelog entry.

**Pull request merged**